IdentityCore/src/requests/broker/MSIDBrokerKeyProvider.m

```diff
@@ -209,4 +209,70 @@ - (BOOL)deleteSymmetricKeyWithError:(NSError **)error
     return YES;
 }
 
+- (BOOL)saveApplicationToken:(NSString *)appToken forClientId:(NSString *)clientId error:(NSError **)error
+{
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"Saving broker application token for clientId %@.", clientId);
+    NSString *tag = [NSString stringWithFormat:@"%@-%@", MSID_BROKER_APPLICATION_TOKEN_TAG, clientId];
+    
+    NSMutableDictionary *applicationTokenAttributes = [NSMutableDictionary new];
+    [applicationTokenAttributes setObject:(id)kSecClassKey forKey:(id)kSecClass];
+    [applicationTokenAttributes setObject:[tag dataUsingEncoding:NSUTF8StringEncoding] forKey:(id)kSecAttrApplicationTag];
+    [applicationTokenAttributes setObject:self.keychainAccessGroup forKey:(id)kSecAttrAccessGroup];
+    [applicationTokenAttributes setObject:(id)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly forKey:(id)kSecAttrAccessible];
+    
+    NSMutableDictionary *update = [NSMutableDictionary dictionary];
+    update[(id)kSecValueData] = [appToken dataUsingEncoding:NSUTF8StringEncoding];
+    
+    OSStatus status = SecItemUpdate((CFDictionaryRef)applicationTokenAttributes, (CFDictionaryRef)update);
+    
+    if (status == errSecItemNotFound)
+    {
+        [applicationTokenAttributes addEntriesFromDictionary:update];
+        status = SecItemAdd((CFDictionaryRef)applicationTokenAttributes, NULL);
+    }
+    
+    if (status != errSecSuccess)
+    {
+        NSString *descr = [NSString stringWithFormat:@"Could not write broker application token %ld.", (long)status];
+        MSIDFillAndLogError(error, MSIDErrorBrokerApplicationTokenWriteFailed, descr, nil);
+        return NO;
+    }
+    
+    return YES;
+}
+
+- (NSString *)getApplicationToken:(NSString *)clientId error:(NSError **)error
+{
+    NSString *tag = [NSString stringWithFormat:@"%@-%@", MSID_BROKER_APPLICATION_TOKEN_TAG, clientId];
+    
+    NSDictionary *applicationTokenQuery =
+    @{
+      (id)kSecClass : (id)kSecClassKey,
+      (id)kSecAttrApplicationTag : [tag dataUsingEncoding:NSUTF8StringEncoding],
+      (id)kSecReturnData : @(YES),
+      (id)kSecAttrAccessGroup : self.keychainAccessGroup
+      };
+    
+    // Get the key bits.
+    CFDataRef applicationToken = nil;
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)applicationTokenQuery, (CFTypeRef *)&applicationToken);
+    
+    if (status != errSecSuccess)
+    {
+        if (status != errSecItemNotFound)
```

Comment:
nit: can we move out this if of parent's if? It would be easier to read.

Reply:
Addressed.